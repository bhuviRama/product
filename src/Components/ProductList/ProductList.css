.avatar {
    vertical-align: middle;
    width: 250px;
    height: 250px;
  }
  span {
    margin: 3%;
  }
  li {
    margin:10%;
    padding: 5%;
  }
  img{
    width: 150px;
    height: 150px;
  }
  .App-header {
    background-color: #e0e0e052;
    /* min-height: 100vh; */
    display: flex;
    flex-direction: row;
    justify-content: center;
    font-size: calc(10px + 2vmin);
  }
  .App-header img{
    width: 100px;
    height: 100px;
    margin-top: 20px;
    padding: 20px;
  }

  .App-header span{
    font-size: 15px;
    text-align: center;
  }
  .most-viewed{
    margin:0;
    align-self: flex-start;
    -moz-transform: translateX(100%);
    -webkit-transform: translateX(100%);
    transform: translateX(100%);
    
    -moz-animation: my-animation 15s linear infinite;
    -webkit-animation: my-animation 15s linear infinite;
    animation: my-animation 15s linear infinite;

  }
  /* for Firefox */
@-moz-keyframes my-animation {
    from { -moz-transform: translateX(-100%); }
    to { -moz-transform: translateX(100%); }
  }
  
  /* for Chrome */
  @-webkit-keyframes my-animation {
    from { -webkit-transform: translateX(-100%); }
    to { -webkit-transform: translateX(100%); }
  }
  
  @keyframes my-animation {
    from {
      -moz-transform: translateX(-100%);
      -webkit-transform: translateX(-100%);
      transform: translateX(-100%);
    }
    to {
      -moz-transform: translateX(100%);
      -webkit-transform: translateX(100%);
      transform: translateX(100%);
    }
  }