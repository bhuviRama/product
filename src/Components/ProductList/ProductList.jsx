import React, { useEffect,useState,useCallback,useMemo } from 'react';
import Typography from '@mui/material/Typography';
import {Link} from "react-router-dom";
import './ProductList.css'   

const ProductList = () => {
    const url = "https://dummyjson.com/products";
    const limit = 10;
    const [ products,setProducts ] = useState([]);
    const [ mostViewedProducts,setMostViewedProducts ] = useState([]);
    let count = 0;
    const getProducts = async() => {
        
        await fetch(url)
        .then((res)=>res.json())
        .then((data)=>{
           
            setProducts(data.products)
        })
    }

    const  getMostViewedProducts = useCallback(() => {
        let arr=[];
        products.map((product)=>{
            let count = window.localStorage.getItem(product.id);
            if(count>0){
                arr.push({
                    id: product.id,
                    views: count,
                })
            }
        })
        console.log(arr)
        arr.sort((a,b) => parseInt(a.views) < parseInt(b.views) ? 1:-1);
        let items = products.filter(({ id: id1 }) => arr.some(({ id: id2 }) => id2 === id1));
        let viewedProducts = [];
        arr.filter((x) => items.map((y) =>{
            if(x.id === y.id){
                viewedProducts.push(y);
            }
        }))
        setMostViewedProducts(viewedProducts);
    },[products])

    useEffect(()=>{
        
        getProducts();
        products.map((product)=>{
            window.localStorage.setItem(product.id, count);
        });
         getMostViewedProducts();
        console.log(mostViewedProducts)
    },[])

    useEffect(()=>{
        getMostViewedProducts();
    },[getMostViewedProducts])

    
//className="App-header"
    return (
        <div>
            {mostViewedProducts.length >0 && 
            <header className="App-header">
                 {/* <Typography align="left" sx={{ fontSize: 14 }} color="text.secondary" >
                    Most viewed
                </Typography> */}
                <span className='most-viewed'>Most viewd products...</span>
                {
                    mostViewedProducts.map((product,index)=>{
                        if(index<3){
                            return(
                                <div >
                                    <Link to={`/products/${product.brand}/${product.title}`} state={{ product}}>
                                        <img  src={product.thumbnail} alt="product thumnail"/>
                                    </Link>
                                    <div style={{ display: 'flex', flexDirection: 'column' }}>
                                        <span>
                                        {product.title}
                                        </span>
                                    </div>
                                </div>
                            )
                        }
                        
                    })
                }
            </header>
}
<Typography variant="h3" gutterBottom>
                    Products
                </Typography>
           {products.length>0 && 
           products.map((product,index)=>{
            if(index<limit ){
                return (
                    <div style={{ display: 'inline-block', padding: '5%' }}>
                <Link to={`/products/${product.brand}/${product.title}`} state={{ product}}>
                    <img className="avatar" src={product.thumbnail} alt="product thumnail"/>
                </Link>
              <div style={{ display: 'flex', flexDirection: 'column' }}>
                <span>
                  {product.title}
                </span>
              </div>
          </div>
                    
                )
            }
            
           })
           }
        </div>
    )
}

export default ProductList;