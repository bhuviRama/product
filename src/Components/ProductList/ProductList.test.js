import React from 'react';
import { create  } from 'react-test-renderer';
import {render, screen} from '@testing-library/react'
import ProductList from './ProductList';

describe('ProductList related tests', () => {
    const fetchMock = jest
    .spyOn(global, 'fetch')
    .mockImplementation(() => Promise.resolve({ json: () => Promise.resolve([]) }))

  test('should render component <ProductList /> properly', () => {
  const component = create(
    <ProductList />
  );

  expect(component.toJSON()).toMatchSnapshot();
  });

  
  it('should fetch all the products', async () => {  
 
    const fakeProducts = [{
        brand:"Apple", 
        category:  "smartphones",
        description :"An apple mobile which is nothing like apple", 
        discountPercentage :12.96, 
        id :1, 
        images:  ['https://i.dummyjson.com/data/products/1/1.jpg', 'https://i.dummyjson.com/data/products/1/2.jpg', 'https://i.dummyjson.com/data/products/1/3.jpg', 'https://i.dummyjson.com/data/products/1/4.jpg', 'https://i.dummyjson.com/data/products/1/thumbnail.jpg'],
        price: 549,
        rating: 4.69,
        stock: 94,
        thumbnail: "https://i.dummyjson.com/data/products/1/thumbnail.jpg",
        title:"iPhone 9"
        }]
       

    fetchMock.mockResolvedValue({ status: 200, json: jest.fn(() => fakeProducts) })
    render (<ProductList />)
    expect(await screen.findByText('Products')).toBeInTheDocument();
   // expect(await screen.findByText('iPhone 9')).toBeInTheDocument()
  })
})


  
