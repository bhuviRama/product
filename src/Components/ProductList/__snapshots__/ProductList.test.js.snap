// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductList related tests should render component <ProductList /> properly 1`] = `
<div>
  <h3
    className="MuiTypography-root MuiTypography-h3 MuiTypography-gutterBottom css-1mktp8m-MuiTypography-root"
  >
    Products
  </h3>
</div>
`;
